proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2245 @@
+long as it is of kind ``Multiplicity``. This is simpler in terms of
+modifying the parser, but the error messages may be confusing for very
+little benefit: in practice we would expect to have polynomial
+expressions of multiplicity variables. Plus, any expression beyond
+this form is unlikely to be resolved by the type checker
+satisfactorily.
+
+So we could decide to restrict ``p`` to the following grammar:
+
+.. code:: bnf
+
+  MULT ::= 'One
+         | 'Omega
+         | VARIABLE
+         | MULT :+ MULT
+         | MULT :* MULT
+         | ( MULT )
+
+Constrained variables
+~~~~~~~~~~~~~~~~~~~~~
+
+Another simple variant on the syntax of ``a :p-> b`` is to restrict
+``p`` to be a variable, and when ``p`` needs to be a composed
+expression, use a constraint of the form ``p ~ q :* r``.
+
+This alternative is probably the simplest in terms of parsing. It has
+the drawback that composed multiplicity expression seem to appear
+mostly in result position. Such as in the composition function
+
+::
+
+  (.) :: (b :q-> c) ->. (a :p-> b) :q-> (a :(p :* q)-> c)
+
+which would become
+
+::
+
+  (.) :: (r ~ p :* q ) => (b :q-> c) ->. (a :p-> b) :q-> (a :r-> c)
+
+It does look a bit curious. But it's a possiblity worth considering.
+
+
+Syntax of the multiplicity-parametric arrow
+-------------------------------------------
+
+The proposed mixfix ``a :p-> b`` syntax for the
+multiplicity-parametric arrow makes a potentially non-trivial addition
+to the parser. So does the proposed type constructor indexed by a
+multiplicity ``(:p->)``.
+
+A way to simplify the changes to the parser would be to have the type
+constructor be
+
+::
+
+  ARROW :: Multiplicity -> * -> * -- ignoring levity
+
+It would be very inconvenient to use a prefix notation for
+multiplicity parametric arrows: we wouldn't want the type of ``map``
+to read
+
+::
+
+  map :: ARROW 'Omega (ARROW p a b) (ARROW p [a] [b])
+
+So we introduce a binary type construction ``WithMult`` (or some
+operator syntax). It is a syntax error to use ``WithMult`` anywhere
+except to the left of an arrow. And ``WithMult a p -> b`` means
+``ARROW p a b``. So that the type of ``map`` becomes:
+
+::
+
+  map :: (a `WithMult` p -> b) -> [a] `WithMult` p -> [b]
+
+Records in GADT syntax
+----------------------
+
+For record in GADT syntax, we proposed that the arrow symbol always be
+``->``, but has no interpretation.
+
+An alternative would be to allow an arbitrary arrow ``:π->`` as in
+
+::
+
+  data R where
+    R :: { f1 ::('One) A, f2 :: B, f3 ::('Omega) C } :π-> R
+
+Which could be interpreted in one of two ways:
+
+- ``π`` can act as a default multiplicity for the fields which don't
+  have a multiplicity annotation. In this case, the type of ``R``
+  would be
+
+  ::
+
+    R :: A ->. B :π-> C -> R
+
+- ``π`` can act as a multiplier on all the fields (unannotated field
+  are considered linear). In this case, the type of ``R`` would be
+
+  ::
+
+    R :: A :π-> B :π-> C -> R
+
+Unboxed data types
```

Comment:
10. This section does not seem to treat unboxed types generally. For example, what does it mean to have

```
f :: Int# ->. ...
f x = ...
```

? Is there any difference to boxed types? Is there a definition of "consumed exactly once" that can apply here?

Reply:
I see, I had misunderstood your question. The proposal treats unboxed types and boxed types identically. But, of course, every primitive type comes with its own definition of consuming exactly once. In the case of unboxed ints, there is not much better we can do besides saying that an unboxed int is consumed exactly once when it is passed to a primitive function.

Fully unboxed types (that is, without any heap reference anywhere in any value), like `Int#` are somewhat special, from the point of view of linearity, in that they are passed by copy. Which means, that, not only can you always implement `move :: Int# ->. Mult# @'Omega Int#` (for the `Mult#` defined just below), but this operation doesn't even emit any code! Therefore linear functions `Int# ->. A` and unrestricted functions `Int# -> A` are essentially the same.

Reply:
OK. It would be nice to clarify this in the proposal (especially since unboxed types don't quite meet the definition of "consumed exactly once", given that they aren't evaluated in the same way as boxed type), but I'm happy with this answer.

Reply:
One could say in general that primitive (unboxed) types (such as Int#, but not ByteArray#) are the least interesting from a linearity point of view, because they are trivial to copy and discard.

Reply:
If I understand correctly, this should mean that the following should typecheck:

    foo :: Int# ->. Int#
    foo x = x +# x

    bar :: Int# ->. Int#
    bar _ = 5#

> every primitive type comes with its own definition of consuming exactly once

Here are the definitions that I understand:

- VecRep VecCount VecElem: always linear consumption
- TupleRep [RuntimeRep]: same rules as lifted data types (pattern match required to be consumed linearly)
- SumRep [RuntimeRep]: same rules as lifted data types (pattern match required to be consumed linearly)
- LiftedRep: normal rules defined in paper
- UnliftedRep: must be used exactly once to be consumed linearly
- IntRep: always linear consumption
- WordRep: always linear consumption
- Int64Rep: always linear consumption
- Word64Rep: always linear consumption
- AddrRep: always linear consumption
- FloatRep: always linear consumption
- DoubleRep: always linear consumption

Do these match what you have in mind?

Reply:
What about `State#` tokens? These have kind `TupleRep '[]`. According to the consumption rules I assume in my previous comment, it would be necessary to pattern match on a value of type `State# s` to consume it linearly, but (fortunately), this isn't possible. For me, this is a practical concern. Ideally, `writeArray#` would have the type

    writeArray# :: MutableArray# s a -> Int# -> a -> State# s ->. State# s

The multiplicity on the arrows for the array and the element must be omega. The mulitplicity on the arrow for the index doesn't matter since it's `Int#`. But, the state token should really be consumed linearly. I sometimes end up in situations writing high-performance code where I have to use these primitives directly, and it would be really nice if the type checker could help me out with ensuring that I consume state tokens linearly. It seems like it can. There's nothing in the proposal about changing the types of anything in `GHC.Prim`, and I don't expect that this would be done initially, but it could be a helpful future addition that should have no impact on people who don't use the extension.

Reply:
> If I understand correctly, this should mean that the following should typecheck:
> 
> foo :: Int# ->. Int#
> foo x = x +# x
>
> bar :: Int# ->. Int#
> bar _ = 5#

We could pull this off (if you are familiar with Rust's `Copy` trait, it would be a similar idea). But I don't think it would be particularly worth the extra complexity in the desugarer. Instead, I meant that you can always do this:

```haskell
foo :: Int# ->. Int#
foo x = case moveInt# x with Mult# x' -> wfoo x'
  where
    wfoo' x = x +# x

bar :: Int# ->. Int#
bar x = case moveInt# x with Mult# x' -> wbar x'
  where
    wbar' _ = 5#
```

`moveInt#` is, in this circumstance, a primitive function (which is actually just a type coercion).
---

For the list of type representations, it look essentially correct.

---

For `State#`, remember that the full characterisation of linear function is: _if_ the returned value is consumed exactly once, _then_ the argument is consumed exactly once. (hence the identity function is linear). In particular, the type you give to `writeArray#` is perfectly acceptable (we argue in that sense somewhere in the paper, I believe; it's about `IO`, but it makes no difference). The only function really capable of (linearly) consuming a `State#` token is `runST`, all the other function must basically pass it along (magically modified by the might of the mind, as is the tradition with `State#` tokens). The linker also kind of does, when it consumes the `main` function.

> I sometimes end up in situations writing high-performance code where I have to use these primitives directly, and it would be really nice if the type checker could help me out with ensuring that I consume state tokens linearly.

I'm actually interested to hear that. Because, to be honest, we were wondering whether this particular tidbit would be useful to anybody, or if the only explicit use of `State#` tokens was in the handful of functions that base export.